pyrobosim/pyrobosim/manipulation/grasping.py

```diff
@@ -0,0 +1,547 @@
+                                           0.0])
+        v_robot_rt_object_proj /= np.linalg.norm(v_robot_rt_object_proj)
+        all_grasp_faces = [GraspFace.FRONT, GraspFace.BACK,
+                           GraspFace.TOP, GraspFace.BOTTOM,
+                           GraspFace.LEFT, GraspFace.RIGHT]
+        max_dot_prod = -10  # Unrealistic value for dot product
+        for face in all_grasp_faces:
+            dot_prod = np.dot(v_robot_rt_object_proj, normal_from_face[face])
+            if dot_prod > max_dot_prod:
+                max_dot_prod = dot_prod
+                front_face_dir = face
+
+        # Figure out what the top face is
+        unit_z = np.dot(
+            tform_robot_to_object[:3, :3].T,
+            np.array([[0.0], [0.0], [1.0]])).T
+        max_dot_prod = -10  # Unrealistic value for dot product
+        for face in all_grasp_faces:
+            dot_prod = np.dot(unit_z, normal_from_face[face])
+            if dot_prod > max_dot_prod:
+                max_dot_prod = dot_prod
+                top_face_dir = face
+
+        # Compute the transform from nominal coordinates to robot-facing
+        x_vec = -1.0 * normal_from_face[front_face_dir]
+        z_vec = normal_from_face[top_face_dir]
+        y_vec = np.cross(-1.0 * x_vec, z_vec)
+        rot_matrix[:,0] = x_vec
+        rot_matrix[:,1] = y_vec
+        rot_matrix[:,2] = z_vec
+        return rot_matrix
+
+    def should_try_grasp(self, faces_enabled, face_normals, face_vec):
+        """
+        Helper function to validate whether to compute grasps on a specific face.
+        
+        :param faces_enabled: Faces enabled, in the form (front, top, side)
+        :type faces_enabled: list[bool]
+        :param face_normals: Face normals in the canonical directions, in the form (front, top, left, right)
+        :type face_normals: list[:class:`numpy.ndarray`]
+        :param face_vec: Face vector to check against
+        :type face_vec: :class:`numpy.ndarray`
+        :return: A tuple determining whether the grasp should be attempted, and what face that corresponds to
+        :type face_vec: (bool, :class:`pyrobosim.manipulation.grasping.GraspFace`)
+        """
+        try_grasp = False
+        grasp_face = GraspFace.UNKNOWN
+
+        front_grasps, top_grasps, side_grasps = faces_enabled
+        front_face_vec, top_face_vec, left_face_vec, right_face_vec = face_normals
+
+        if front_grasps and np.allclose(front_face_vec, face_vec):
+            try_grasp = True
+            grasp_face = GraspFace.FRONT
+        elif top_grasps and np.allclose(top_face_vec, face_vec):
+            try_grasp = True
+            grasp_face = GraspFace.TOP
+        elif side_grasps:
+            if np.allclose(left_face_vec, face_vec):
+                try_grasp = True
+                grasp_face = GraspFace.LEFT
+            elif np.allclose(right_face_vec, face_vec):
+                try_grasp = True
+                grasp_face = GraspFace.RIGHT
```

Comment:
Prob could improve this by also checking if (front_face_vec_width > gripper_max_width), for example. Because there's no use trying a grasp if the fingers can't open wide enough.

I guess you would need a ParallelGraspProperties input to this function, then.

Reply:
Yes, but we already do that filtering in the `generate()` function outside this, e.g.,

```python
if effective_max_width >= object_y:
```